### PR TITLE
ConsoleTarget - Apply Encoding on InitializeTarget, if Console available

### LINKED ISSUE #1845: Console target - encoding throws IOException
**Type** : Bug
**NLog version**: 4.3.11
**Platform**: .Net 4.6.2

- I set the console encoding to UTF-8 and got IOException when console is not available (ex: running app as service or running integration tests - mstest's [without interactive mode]).
- As a workaround - when encoding was not set everything works fine, detectConsoleAvailable="true" is not necessary.
- Bug appears on console and coloredconsole target type. 

- **app.config:**

```xml
<configuration>
  <configSections>
    ...
    <section name="nlog" type="NLog.Config.ConfigSectionHandler, NLog" />
  </configSections>
  <nlog>
    <include file="${basedir}/Configuration/Environment/NLog.config" />
  </nlog>
</configuration>
```

- **NLog.config:**

```xml
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
      xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
      autoReload="true"
      throwExceptions="true"
      internalLogLevel="Debug"
      internalLogFile="c:\temp\nlog-internal.log">

  <targets>
    <target name="console" xsi:type="ColoredConsole"
            layout="${level:uppercase=true}@${longdate}@${gdc:item=node}@: ${message} ${exception:format=tostring}"
            encoding="UTF-8"
            detectConsoleAvailable="true"/>
  </targets>

  <rules>
    <logger minlevel="Warn" writeTo="console" name="*" />
  </rules>
</nlog>
```

- **I've checked internal log and there is exception (from internal log):**

```
2016-12-12 10:55:39.4746 Debug Setting 'ColoredConsoleTarget.encoding' to 'UTF-8'
2016-12-12 10:55:39.4746 Error Error when including 'C:\SampleNLogIOExceptionApp\/Configuration/Environment/NLog.config'. Exception: NLog.NLogConfigurationException: Error when setting property 'Encoding' on ColoredConsole Target[console] ---> System.IO.IOException: Nieprawidłowe dojście.

   w System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   w System.Console.set_OutputEncoding(Encoding value)
   --- Koniec śladu stosu wyjątków wewnętrznych ---
   w NLog.Internal.PropertyHelper.SetPropertyFromString(Object obj, String propertyName, String value, ConfigurationItemFactory configurationItemFactory)
   w NLog.Config.XmlLoggingConfiguration.ConfigureObjectFromAttributes(Object targetObject, NLogXmlElement element, Boolean ignoreType)
   w NLog.Config.XmlLoggingConfiguration.ParseTargetElement(Target target, NLogXmlElement targetElement)
   w NLog.Config.XmlLoggingConfiguration.ParseTargetsElement(NLogXmlElement targetsElement)
   w NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ConfigureFromFile(String fileName, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseIncludeElement(NLogXmlElement includeElement, String baseDirectory, Boolean autoReloadDefault)
2016-12-12 10:55:39.4866 Error Error in Parsing Configuration File. Exception: NLog.NLogConfigurationException: Exception occurred when loading configuration from C:\SampleNLogIOExceptionApp\SampleNLogIOExceptionApp.exe.Config ---> NLog.NLogConfigurationException: Error when setting property 'Encoding' on ColoredConsole Target[console] ---> System.IO.IOException: Nieprawidłowe dojście.

   w System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   w System.Console.set_OutputEncoding(Encoding value)
   --- Koniec śladu stosu wyjątków wewnętrznych ---
   w NLog.Internal.PropertyHelper.SetPropertyFromString(Object obj, String propertyName, String value, ConfigurationItemFactory configurationItemFactory)
   w NLog.Config.XmlLoggingConfiguration.ConfigureObjectFromAttributes(Object targetObject, NLogXmlElement element, Boolean ignoreType)
   w NLog.Config.XmlLoggingConfiguration.ParseTargetElement(Target target, NLogXmlElement targetElement)
   w NLog.Config.XmlLoggingConfiguration.ParseTargetsElement(NLogXmlElement targetsElement)
   w NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ConfigureFromFile(String fileName, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseIncludeElement(NLogXmlElement includeElement, String baseDirectory, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.Initialize(XmlReader reader, String fileName, Boolean ignoreErrors)
   --- Koniec śladu stosu wyjątków wewnętrznych ---
2016-12-12 10:55:39.4866 Error ConfigSectionHandler error. Exception: NLog.NLogConfigurationException: Exception occurred when loading configuration from C:\SampleNLogIOExceptionApp\SampleNLogIOExceptionApp.exe.Config ---> NLog.NLogConfigurationException: Error when setting property 'Encoding' on ColoredConsole Target[console] ---> System.IO.IOException: Nieprawidłowe dojście.

   w System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   w System.Console.set_OutputEncoding(Encoding value)
   --- Koniec śladu stosu wyjątków wewnętrznych ---
   w NLog.Internal.PropertyHelper.SetPropertyFromString(Object obj, String propertyName, String value, ConfigurationItemFactory configurationItemFactory)
   w NLog.Config.XmlLoggingConfiguration.ConfigureObjectFromAttributes(Object targetObject, NLogXmlElement element, Boolean ignoreType)
   w NLog.Config.XmlLoggingConfiguration.ParseTargetElement(Target target, NLogXmlElement targetElement)
   w NLog.Config.XmlLoggingConfiguration.ParseTargetsElement(NLogXmlElement targetsElement)
   w NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ConfigureFromFile(String fileName, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseIncludeElement(NLogXmlElement includeElement, String baseDirectory, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseNLogElement(NLogXmlElement nlogElement, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.ParseTopLevel(NLogXmlElement content, String filePath, Boolean autoReloadDefault)
   w NLog.Config.XmlLoggingConfiguration.Initialize(XmlReader reader, String fileName, Boolean ignoreErrors)
   --- Koniec śladu stosu wyjątków wewnętrznych ---
   w NLog.Config.XmlLoggingConfiguration.Initialize(XmlReader reader, String fileName, Boolean ignoreErrors)
   w NLog.Config.XmlLoggingConfiguration..ctor(XmlElement element, String fileName)
   w NLog.Config.ConfigSectionHandler.Create(XmlNode section, IAppDomain appDomain)
2016-12-12 10:55:39.5026 Debug Including file 'C:\SampleNLogIOExceptionApp\/Configuration/Environment/NLog.config'
```

### PR DESCRIPTION
**Pull request by snakefoot**
Attempt to fix #1845 using detectConsoleAvailable=true

---
This change is [Reviewable](https://reviewable.io/reviews/nlog/nlog/1850)

### CODE REVIEW

File: src/NLog/Targets/ConsoleTarget.cs [hunk 1]
```diff
@@ -86,17 +86,23 @@ public sealed class ConsoleTarget : TargetWithLayoutHeaderAndFooter
         [DefaultValue(false)]
         public bool Error { get; set; }
 
-#if !SILVERLIGHT && !__IOS__ && !__ANDROID__
         /// <summary>
         /// The encoding for writing messages to the <see cref="Console"/>.
         ///  </summary>
         /// <remarks>Has side effect</remarks>
         public Encoding Encoding
```
Comment (304NotModified) [inline src/NLog/Targets/ConsoleTarget.cs]: In  SILVERLIGHT / __IOS__  / __ANDROID__ we have now the property `Encoding`, but it doesn't work. Do you think that's an improvement?

Comment (snakefoot) [inline src/NLog/Targets/ConsoleTarget.cs]: I have no opinion. Just removed the defines, since the property was no longer using the console directly.

**Comment by snakefoot** [comment]
I guess one could try and test the encoding, but I can see you added with encoding property with any tests :)

**Comment by snakefoot** [comment]
Made a unit-test, and found a bug when dynamically changing ConsoleTarget.Encoding after it has been initialized. Guess I learned my lesson once again :)

File: src/NLog/Targets/ConsoleTarget.cs (hunk 1 above)
Comment (304NotModified) [inline src/NLog/Targets/ConsoleTarget.cs]: I prefer removing it, as it won't do anything. This could be confusing from the API.

**Comment by 304NotModified** [comment]
Looks good, only the API change is a breaker (for SILVERLIGHT /__IOS__ /__ANDROID__).

**Comment by 304NotModified** [comment]
Thanks!

**Pull request merged**